funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/task_queue_subscriber.py

```diff
@@ -0,0 +1,401 @@
+        else:
+            self.acknowledge_message(basic_deliver.delivery_tag)
+
+    def acknowledge_message(self, delivery_tag):
+        """Acknowledge the message delivery from RabbitMQ by sending a
+        Basic.Ack RPC method for the delivery tag.
+
+        :param int delivery_tag: The delivery tag from the Basic.Deliver frame
+
+        """
+        logger.info(f"Acknowledging message {delivery_tag}")
+        self._channel.basic_ack(delivery_tag)
+
+    def stop_consuming(self):
+        """Tell RabbitMQ that you would like to stop consuming by sending the
+        Basic.Cancel RPC command.
+
+        """
+        if self._channel:
+            logger.info("Sending a Basic.Cancel RPC command to RabbitMQ")
+            self._channel.basic_cancel(self.on_cancelok, self._consumer_tag)
+
+    def on_cancelok(self, unused_frame):
+        """This method is invoked by pika when RabbitMQ acknowledges the
+        cancellation of a consumer. At this point we will close the channel.
+        This will invoke the on_channel_closed method once the channel has been
+        closed, which will in-turn close the connection.
+
+        :param pika.frame.Method unused_frame: The Basic.CancelOk frame
+
+        """
+        logger.info("RabbitMQ acknowledged the cancellation of the consumer")
+        self.close_channel()
+
+    def close_channel(self):
+        """Call to close the channel with RabbitMQ cleanly by issuing the
+        Channel.Close RPC command.
+
+        """
+        logger.info("Closing the channel")
+        self._channel.close()
+
+    def event_watcher(self):
+        """Polls the kill_event periodically to trigger a shutdown"""
+        if self.kill_event.is_set():
+            logger.info("Kill event is set. Start subscriber shutdown")
+            try:
+                self.close_connection()
+                self._connection.ioloop.stop()
+                self.cleanup_complete.set()
+            except Exception:
+                logger.exception("Failed to close connection cleanly")
+            logger.info("Shutdown complete")
+            return
+        else:
+            self._connection.ioloop.call_later(
+                self._watcher_poll_period, self.event_watcher
+            )
+
+    def run(self):
+        """Run the example consumer by connecting to RabbitMQ and then
+        starting the IOLoop to block and allow the SelectConnection to
+        operate.
+
+        Note: Only one of these options should be used.
+        """
+        logger.warning("Run method")
+        try:
+            self._connection = self._connect()
+            self.event_watcher()
+            self._connection.ioloop.start()
+        except Exception:
+            logger.exception("Failed to start subscriber")
+
+    def close(self) -> None:
+        """This close method needs some rethinking since this whole thing is a separate
+        process with the ioloop callbacks not setting the vars properly
+        """
```

Comment:
Given that `Process.close` is already a thing, I don't think we should override it without invoking `super().close()` at the end. However, if this needs to be rethought, perhaps the better approach is to remove this method altogether.

If you remove `close()` and `cleanup_complete` from this class, then setting the kill event would potentially work as the only shutdown mechanism. I haven't carefully thought through the details, but it's worth considering.

At the very, very least, this would need to call `super().close()` to behave correctly.

Reply:
I've updated this to do something like this:
```
# use kill_event to get the poller to shutdown the eventloop 
self.join()
super().close()
```

Reply:
I'm not seeing that edit yet -- not sure what's up exactly.

I'm also looking at `self._closing = True` as suspicious. `close()` is meant to be called by the parent process, and `_closing` is being used for managing state in the child.
More generally, I think there's a very undesirable mixing happening right now because the subscriber object _is_ the Process  class. It might be possible to fix this by having
```python
class TaskQueueSubscriber: ...
class TaskQueueSubscriberProc(Process): ...
```
with some attribute in the "proc" class containing the "real subscriber" object.

Let's revisit that idea later. For now, I just want to make sure `close()` either works (from the parent) or that we don't use it and instead we have a clear termination pattern from the parent like `subproc.set_kill_event(); subproc.join()`.

Reply:
Looks good to me with the change. I'll read again to see if `self._closing` should be set in the parent process.